### DISCUSSION TITLE
Source term calculation

### DISCUSSION THREAD

**Discussion by styyokuda** [idaholab/moose#27721] [Q&A General] [answered]
### Check these boxes if you have followed the posting rules.

- [X] Q&A General is the most appropriate section for my question
- [X] I have consulted the posting Guidelines on the Discussions front page
- [X] I have searched the Discussions forum and my question has not been asked before
- [X] I have searched the MOOSE website and the documentation does not answer my question
- [X] I have formatted my post following the posting guidelines (screenshots as a last resort, triple back quotes around pasted text)

### Question

All,

In my model, I solve diffusion and heat equations. For the diffusion equations, I need to calculate the source/sink terms which are dependent on variables and parameters and it needs more than the "functions" with "ParsedFunction" type calculations. Please let me know if I should use the "Material system computing" or something else?
Please help. S. Thomas

**Comment by GiudGiud** [comment, marked as answer]
Hello

If the reaction / source / sink terms are simple linear functions of a variable then you can use CoupledForce to add that term to an equation

If it s any more complicated then I would recommend using ADMatCoupledForce with an ADParsedMaterial to describe the term

Guillaume

**Reply by styyokuda** [reply]
All,

I would like to use the sink value I calculate in "Material System" and input into "BodyForce" in "Kernels System" via "Functions System."  In the attached function object "SinkMat", "_value_mat[_qp]" is the sink value I calculate in "Material System."  I got an error message for "_value_mat(getMaterialProperty<Real>("value_mat"))."  Is this approach not work?  Please help.  S. Thoms
[SinkMat.pdf](https://github.com/user-attachments/files/16089297/SinkMat.pdf)
[SinkMat.pdf](https://github.com/user-attachments/files/16089299/SinkMat.pdf)

**Reply by styyokuda** [reply]
All,
I would like to make sure that I put the correct files in the previous post.
Please help, S. Thomas
[SinkMat_C.pdf](https://github.com/user-attachments/files/16089358/SinkMat_C.pdf)
[SinkMat_h.pdf](https://github.com/user-attachments/files/16089361/SinkMat_h.pdf)

**Reply by styyokuda** [reply]
All,
I also tried to modify the BodyForce" object (please see Attached).  It is not still working.
Pease help, S. Thomas
[BodyForceMat_C.pdf](https://github.com/user-attachments/files/16090668/BodyForceMat_C.pdf)
[BodyForceMat_h.pdf](https://github.com/user-attachments/files/16090672/BodyForceMat_h.pdf)

**Reply by GiudGiud** [reply]
>  I calculate in "Material System" and input into "BodyForce" in "Kernels System" via "Functions System."

you cant use Materials in Functors.
You need to find a different kernel. Or create a new one

For example CoupledForce can use a variable, which can bring material properties through the ParsedAux auxkernel

**Reply by GiudGiud** [reply]
for your first comment, I think the error is that you are missing the parameter in the validParams
you should have
```
 InputParameters params = Function::validParams();
 params.addClassDescription("Provide the sink value");
  params.addParam<MaterialPropertyName>("value_mat", <a default if it makes sense, if not then none>, "description");
 return params;

**Reply by styyokuda** [reply]
Thank very much for the suggestion. I still have errors as below which I don't understand.
Pease help, S. Thomas

```
Compiling C++ (in opt mode) /Users/768/projects/run/build/unity_src/functions_Unity.C...
In file included from /Users/768/projects/run/build/unity_src/functions_Unity.C:2:
/Users/768/projects/run/src/functions/SinkMat.C:19:54: error: expected expression
  params.addParam<MaterialPropertyName>("value_mat", <a default if it makes sense, if not then none>, "description");
                                                     ^
/Users/768/projects/run/src/functions/SinkMat.C:19:55: error: use of undeclared identifier 'a'
  params.addParam<MaterialPropertyName>("value_mat", <a default if it makes sense, if not then none>, "description");
                                                      ^
/Users/768/projects/run/src/functions/SinkMat.C:19:84: error: expected expression
  params.addParam<MaterialPropertyName>("value_mat", <a default if it makes sense, if not then none>, "description");
                                                                                   ^
/Users/768/projects/run/src/functions/SinkMat.C:25:16: error: use of undeclared identifier 'getMaterialProperty'
    _value_mat(getMaterialProperty<Real>("value_mat"))
               ^
/Users/768/projects/run/src/functions/SinkMat.C:32:21: error: use of undeclared identifier '_qp'
  return _value_mat[_qp];
```

**Reply by GiudGiud** [reply]
hello

when we `<...>` in code it's meant for you to fill in, you can't copy paste it in it wont work without filling it

**Reply by styyokuda** [reply]
Sorry ... I will type then.  The errors are:
.../functions_Unity.C:2:: ... /functions/SinkMat.C:19:54: error: expected expression.

`
params.addParam<MaterialPropertyName>("value_mat",<a default if it makes sense, if not then none>, "description");
`

... /functions/SinkMat.C:19:55: error: use of undeclared identifier 'a'

`
params.addParam<MaterialPropertyName>("value_mat", <a default if it makes sense, if not then none>, "description");
`

... /functions/SinkMat.C:25:16: error: use of undeclared identifier 'getMaterialProperty'
_value_mat(getMaterialProperty<Real>('value_mat")

**Reply by GiudGiud** [reply]
```
params.addParam<MaterialPropertyName>("value_mat", <a default if it makes sense, if not then none>, "description");
```

Replace `<a default if it makes sense, if not then none>` with a a default
If you dont understand what I mean, replace with this, which gives a 0 default
```
params.addParam<MaterialPropertyName>("value_mat", "0",  "description");

```

**Reply by styyokuda** [reply]
I still receive an error of
```
... /functions/SinkMat.C:25:16: error: use of undeclared identifier 'getMaterialProperty'
_value_mat(getMaterialProperty('value_mat")
```
Using the ParsedAux auxkernel as you suggested before may be a good idea.
S. Thomas

**Reply by GiudGiud** [reply]
Because this is not the correct syntax. It should be something like
```
_value_mat(getMaterialProperty<Real>("value_mat"))
```

**Reply by styyokuda** [reply]
Sorry, it was my mistype. Correctly it was:
`
_value_mat(getMaterialProperty<Real>("value_mat"))
`
Thank you, S, Thomas

**Reply by styyokuda** [reply]
Thank you very much, Guillaume!

Your suggestion of using CoupledForce with the ParsedAux auxkernel works well.  :)
I appreciate your great help, as always.
S. Thomas
